src/main/java/org/javarosa/xpath/expr/XPathSortFunc.java

```diff
@@ -0,0 +1,146 @@
+        super(NAME, args, EXPECTED_ARG_COUNT, true);
+    }
+
+    @Override
+    protected void validateArgCount() throws XPathSyntaxException {
+        if (args.length < 1 || args.length > 3) {
+            throw new XPathArityException(name, "between 1 and 3 arguments", args.length);
+        }
+    }
+
+    @Override
+    protected Object evalBody(DataInstance model, EvaluationContext evalContext, Object[] evaluatedArgs) {
+        List<String> sortedList;
+        if (evaluatedArgs.length == 1) {
+            sortedList = sortSingleList(FunctionUtils.toString(evaluatedArgs[0]), true);
+        } else if (evaluatedArgs.length == 2) {
+            Object lastArg = evaluatedArgs[1];
+            if (lastArg instanceof Boolean) {
+                sortedList =
+                        sortSingleList(FunctionUtils.toString(evaluatedArgs[0]),
+                                FunctionUtils.toBoolean(evaluatedArgs[1]));
+            } else {
+                sortedList = sortListByOtherList(FunctionUtils.toString(evaluatedArgs[0]),
+                        FunctionUtils.toString(evaluatedArgs[1]), true);
+            }
+        } else {
+            sortedList =
+                    sortListByOtherList(FunctionUtils.toString(evaluatedArgs[0]),
+                            FunctionUtils.toString(evaluatedArgs[1]),
+                            FunctionUtils.toBoolean(evaluatedArgs[2]));
+        }
+        return listToString(sortedList);
+    }
+
+    private List<String> sortSingleList(String spaceSeparatedString, boolean ascending) {
+        List<String> items = stringToList(spaceSeparatedString);
+        sortSingleList(items, ascending);
+        return items;
+    }
+
+    private void sortSingleList(List<String> items, final boolean ascending) {
+        Collections.sort(items, new Comparator<String>() {
+            @Override
+            public int compare(String s1, String s2) {
+                return (ascending ? 1 : -1) * s1.compareTo(s2);
+            }
+        });
+    }
+
+    private List<String> sortListByOtherList(String s1, String s2, boolean ascending) {
+        List<String> targetListItems = stringToList(s1);
+        List<String> comparisonListItems = stringToList(s2);
+
+        if (targetListItems.size() != comparisonListItems.size()) {
+            throw new XPathTypeMismatchException("Length of lists passed to sort() must match");
+        }
+
+        Map<String, List<String>> stringMapping =
+                createMappingFromComparisonToTarget(comparisonListItems, targetListItems);
+
+        List<String> sortedComparisonList = sortSingleList(s2, ascending);
+        List<String> sortedTargetList = new ArrayList<>();
+        String previousComparisonString = "";
+        for (int i = 0; i < sortedComparisonList.size(); i++) {
+            String stringInSortedList = sortedComparisonList.get(i);
+            if (stringInSortedList.equals(previousComparisonString)) {
```

Comment:
We may need to sort lists where there are multiple keys on either side of the sorting, and i'm pretty concerned with how this is going to deal with those.

It's pretty annoying to handle a lot of string manipulations, I think the safest option would be if, rather than making the map above, you made a list of tuples [targetListItem, comparisonListItem] (we can use Pair<> in the core code now) and you made a comparator that just sorted by the second member of the tuple. That should handle the case where there are string matching duplicates and I think will likely make the code a bit clearer.